### PR TITLE
InferAddressSpaces: Factor replacement loop into function [NFC]

### PR DESCRIPTION
**Pull request by arsenm**

### CODE REVIEW

**Comment by arsenm** [comment]
* **#104430** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/104430?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @arsenm and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by arsenm** [comment]
ping

File: llvm/lib/Transforms/Scalar/InferAddressSpaces.cpp [hunk 1]
```diff
@@ -842,8 +846,9 @@ unsigned InferAddressSpacesImpl::joinAddressSpaces(unsigned AS1,
   return (AS1 == AS2) ? AS1 : FlatAddrSpace;
 }
 
-bool InferAddressSpacesImpl::run(Function &F) {
-  DL = &F.getDataLayout();
+bool InferAddressSpacesImpl::run(Function &F_) {
+  F = &F_;
```
Comment (Pierre-vh) [inline llvm/lib/Transforms/Scalar/InferAddressSpaces.cpp:850]: Not sure I like the trailing _. Maybe use something like `CurFn` instead?

File: llvm/lib/Transforms/Scalar/InferAddressSpaces.cpp [hunk 2]
```diff
@@ -1164,10 +1169,105 @@ static Value::use_iterator skipToNextUser(Value::use_iterator I,
...
+    if (!MI->isVolatile() && handleMemIntrinsicPtrUse(MI, V, NewV))
+      return;
+  }
+
+  if (auto *II = dyn_cast<IntrinsicInst>(CurUser)) {
+    if (rewriteIntrinsicOperands(II, V, NewV))
+      return;
+  }
+
+  if (isa<Instruction>(CurUser)) {
```
Comment (Pierre-vh) [inline llvm/lib/Transforms/Scalar/InferAddressSpaces.cpp:1201]: Flip condition, early return?
Can also convert it into a `dyn_cast` to Instruction to simplify the `CurUserI` check above too.

**Review by Pierre-vh — COMMENTED** [review]
Add [NFC] ?

File: llvm/lib/Transforms/Scalar/InferAddressSpaces.cpp (hunk 1 above)
Comment (arsenm) [reply llvm/lib/Transforms/Scalar/InferAddressSpaces.cpp]: this convention is followed in maybe half of places? Could also just move setting F to the constructor but that just moves the name collision (to a less bad place, I guess)

Comment (Pierre-vh) [reply llvm/lib/Transforms/Scalar/InferAddressSpaces.cpp]: Using something like `F(&F)` in the constructor list is well defined in C++ and wouldn't be a collision.

In any case, I wasn't aware this was a convention and I'm not going to block anything over this :)

File: llvm/lib/Transforms/Scalar/InferAddressSpaces.cpp [hunk 3]
```diff
@@ -1168,10 +1173,103 @@ static Value::use_iterator skipToNextUser(Value::use_iterator I,
...
+    if (!MI->isVolatile() && handleMemIntrinsicPtrUse(MI, V, NewV))
+      return;
+  }
+
+  if (auto *II = dyn_cast<IntrinsicInst>(CurUser)) {
+    if (rewriteIntrinsicOperands(II, V, NewV))
+      return;
+  }
+
+  if (ICmpInst *Cmp = dyn_cast<ICmpInst>(CurUserI)) {
```
Comment (Pierre-vh) [inline llvm/lib/Transforms/Scalar/InferAddressSpaces.cpp:1205]: nit: could use auto here

File: llvm/lib/Transforms/Scalar/InferAddressSpaces.cpp [hunk 4]
```diff
@@ -1168,10 +1173,103 @@ static Value::use_iterator skipToNextUser(Value::use_iterator I,
...
+      if (isSafeToCastConstAddrSpace(KOtherSrc, NewAS)) {
+        Cmp->setOperand(SrcIdx, NewV);
+        Cmp->setOperand(OtherIdx, ConstantExpr::getAddrSpaceCast(
+                                      KOtherSrc, NewV->getType()));
+        return;
+      }
+    }
+  }
+
+  if (AddrSpaceCastInst *ASC = dyn_cast<AddrSpaceCastInst>(CurUserI)) {
```
Comment (Pierre-vh) [inline llvm/lib/Transforms/Scalar/InferAddressSpaces.cpp:1236]: ditto

File: llvm/lib/Transforms/Scalar/InferAddressSpaces.cpp [hunk 5]
```diff
@@ -1168,10 +1173,103 @@ static Value::use_iterator skipToNextUser(Value::use_iterator I,
...
+    unsigned NewAS = NewV->getType()->getPointerAddressSpace();
+    if (ASC->getDestAddressSpace() == NewAS) {
+      ASC->replaceAllUsesWith(NewV);
+      DeadInstructions.push_back(ASC);
+      return;
+    }
+  }
+
+  // Otherwise, replaces the use with flat(NewV).
+  if (Instruction *VInst = dyn_cast<Instruction>(V)) {
```
Comment (Pierre-vh) [inline llvm/lib/Transforms/Scalar/InferAddressSpaces.cpp:1246]: ditto

File: llvm/lib/Transforms/Scalar/InferAddressSpaces.cpp [hunk 6]
```diff
@@ -1168,10 +1173,103 @@ static Value::use_iterator skipToNextUser(Value::use_iterator I,
...
+    // Don't create a copy of the original addrspacecast.
+    if (U == V && isa<AddrSpaceCastInst>(V))
+      return;
+
+    // Insert the addrspacecast after NewV.
+    BasicBlock::iterator InsertPos;
+    if (Instruction *NewVInst = dyn_cast<Instruction>(NewV))
```
Comment (Pierre-vh) [inline llvm/lib/Transforms/Scalar/InferAddressSpaces.cpp:1253]: ditto

**Review by Pierre-vh — APPROVED** [review]
LGTM, some minor nits

**Comment by arsenm** [comment]
### Merge activity

* **Aug 20, 11:38 AM EDT**: @arsenm started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/104430).

**Pull request merged**